### DISCUSSION TITLE
What is "named content part" and when should we use it?

### DISCUSSION THREAD

**Discussion by aaronamm** [OrchardCMS/OrchardCore#7568] [Q&A]
Hello everyone

I just found we have "named content part" in Orchard Core.
I know that Orchard has had content part since a version 1 and it has a class model along with it.

After digging some code, I still don't quite understand the purpose of "named content part" and there it is not mentioned in OrchardCore doc.

Here are what I founds:
- Article name part in blog.recipe.json
```
              "PartName": "Article",
              "Name": "Article",
              "Settings": {
                "ContentTypePartSettings": {
                  "Position": "3"
                }
```
- In admin content type definition that we can add a named part
![image](https://user-images.githubusercontent.com/2938310/98665041-b6d18280-237d-11eb-8192-82583aeec256.png)

My question are:
- Please could you explain what is named part?
- What is the purpose of it?
- I wonder if a problem we have can be solved with a traditional content part.

Thanks.

FYI 

I can't figure it out the real different between AddPartToType and AddReusablePartToType in "ContentDefinitionService.cs"
``` c#
        public void AddPartToType(string partName, string typeName)
        {
            _contentDefinitionManager.AlterTypeDefinition(typeName, typeBuilder => typeBuilder.WithPart(partName));
            _contentDefinitionEventHandlers.Invoke((handler, context) => handler.ContentPartAttached(context), new ContentPartAttachedContext { ContentTypeName = typeName, ContentPartName = partName }, _logger);
        }

        public void AddReusablePartToType(string name, string displayName, string description, string partName, string typeName)
        {
            _contentDefinitionManager.AlterTypeDefinition(typeName, typeBuilder => typeBuilder.WithPart(name, partName, cfg =>
            {
                cfg.WithDisplayName(displayName);
                cfg.WithDescription(description);
            }));

            _contentDefinitionEventHandlers.Invoke((handler, context) => handler.ContentPartAttached(context), new ContentPartAttachedContext { ContentTypeName = typeName, ContentPartName = partName }, _logger);
        }

```

**Comment by jptissot** [comment]
I think the difference is that usually you can only attach 1 reference of a part to a content item. But for Named Parts, you can attach X instances, each would have a different names. 

For example, the BagPart can be attached multiple times to the same item for different purposes.

**Reply by aaronamm** [reply]
@jptissot Thank you for your answer. I understand what you explained but I still don't quite understand why we need named part `Article` in `Article` content type. Is it possible if you could explain it to me?

Here is how we attach Article named part to Article content type.
https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Themes/TheBlogTheme/Recipes/blog.recipe.json#L113

Thanks.

**Reply by deanmarcussen** [reply]
The `Article` part you are referring to is a special part which contains the fields for the current content type.

It's not exactly a NamedPart, just the part where fields are placed.

NamedParts are by design, parts that be attached to a content type multiple times, with a different name

**Reply by aaronamm** [reply]
@deanmarcussen  Thank you so much.
However, could you help point find the field that `Article` part store?
I cannot find any "Article.cs" model which differs from `AutoroutePart`, `HtmlBodyPart`, `TitlePart` that ways have xxxPart.cs.

Thanks.

**Reply by deanmarcussen** [reply]
There is no Model. It's where fields are placed. Dynamically. Through the admin ui. 

In the case of `Article` in that recipe, it has a `TextField` named Subtitle and `MediaField` on it

**Reply by aaronamm** [reply]
@deanmarcussen Thank you for your prompt reply.
Now, I get it. 

Here is how we define dynamic Article part, just for reference.

Only defining of Article part in `src/OrchardCore.Themes/TheBlogTheme/Recipes/blog.recipe.json`.
```
{
   "ContentParts":[
      {
         "Name":"Article",
         "Settings":{
            
         },
         "ContentPartFieldDefinitionRecords":[
            {
               "FieldName":"TextField",
               "Name":"Subtitle",
               "Settings":{
                  "ContentPartFieldSettings":{
                     "DisplayName":"Subtitle",
                     "Position":"0"
                  }
               }
            },
            {
               "FieldName":"MediaField",
               "Name":"Image",
               "Settings":{
                  "ContentPartFieldSettings":{
                     "DisplayName":"Banner Image",
                     "Position":"0"
                  },
                  "MediaFieldSettings":{
                     "Multiple":false,
                     "AllowAnchors":true
                  }
               }
            }
         ]
      }
   ]
}
```
